src/main-container/private/ActiveMainProvider.jsx

```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import uuidv4 from 'uuid/v4';
+
+import NavigationItemContext from '../../navigation-item/NavigationItemContext';
+
+import ActiveMainContext from '../ActiveMainContext';
+import ActiveMainRegistrationContext from './ActiveMainRegistrationContext';
+
+const propTypes = {
+  children: PropTypes.node,
+};
+
+const ActiveMainProvider = ({ children }) => {
+  const navigationItem = React.useContext(NavigationItemContext);
+  const activeMainRegistration = React.useContext(ActiveMainRegistrationContext);
+  const unregisterActiveMainRef = React.useRef();
+
+  const [state, dispatch] = React.useReducer((currentState, action) => {
+    if (action.type === 'register') {
+      return {
+        registrationId: action.registrationId,
+        activeMain: {
+          label: action.label,
+          metaData: action.metaData,
+        },
+      };
+    }
+
+    if (action.type === 'unregister') {
+      if (currentState.registrationId === action.registrationId) {
+        return {
+          registrationId: undefined,
+          activeMainPage: undefined,
+        };
+      }
+    }
+
+    return currentState;
+  }, { registrationId: undefined, activeMain: undefined });
+
+  React.useEffect(() => {
+    if (!activeMainRegistration) {
+      return;
+    }
+
+    // If an ancestor provider exists, we need to forward the active main info
+    // if the provider exists in the active navigation branch. It is
+    // otherwise unregistered, if necessary, as we do not want potentially stale
+    // information living above this provider level.
```

Comment:
Having nested active main providers should never be a thing right? If that happens maybe we should just blow up?

Reply:
It is a thing, as the ActiveMainProviders are rendered both by the ApplicationContainer as well as by each NavigationItem (which will be inside the ApplicationContainer, and potentially nested even further if we have secondary/tertiary navigation concepts).